programs/drift/src/state/perp_market.rs

```diff
@@ -1686,3 +1695,227 @@ impl AMM {
+}
+
+#[zero_copy]
+#[derive(AnchorSerialize, AnchorDeserialize, Debug)]
+#[repr(C)]
+pub struct CacheInfo {
+    pub last_fee_pool_token_amount: u128,
+    pub last_net_pnl_pool_token_amount: i128,
+    /// BASE PRECISION
+    pub position: i64,
+    pub slot: u64,
+    pub max_confidence_interval_multiplier: u64,
+    pub last_oracle_price_twap: i64,
+    pub last_settle_amount: u64,
+    pub last_settle_slot: u64,
+    pub quote_owed_from_lp_pool: i64,
+    pub oracle_price: i64,
+    pub oracle_confidence: u64,
+    pub oracle_delay: i64,
+    pub oracle_slot: u64,
+    pub oracle_source: u8,
+    pub _padding: [u8; 7],
+    pub oracle: Pubkey,
+}
+
+impl Size for CacheInfo {
+    const SIZE: usize = 160 + 8 + 8 + 8;
+}
+
+impl Default for CacheInfo {
+    fn default() -> Self {
+        CacheInfo {
+            position: 0i64,
+            slot: 0u64,
+            max_confidence_interval_multiplier: 1u64,
+            last_oracle_price_twap: 0i64,
+            oracle_price: 0i64,
+            oracle_confidence: 0u64,
+            oracle_delay: 0i64,
+            oracle_slot: 0u64,
+            _padding: [0u8; 7],
+            oracle: Pubkey::default(),
+            last_fee_pool_token_amount: 0u128,
+            last_net_pnl_pool_token_amount: 0i128,
+            last_settle_amount: 0u64,
+            last_settle_slot: 0u64,
+            oracle_source: 0u8,
+            quote_owed_from_lp_pool: 0i64,
+        }
+    }
+}
+
+impl CacheInfo {
+    pub fn get_oracle_source(&self) -> DriftResult<OracleSource> {
+        Ok(OracleSource::try_from(self.oracle_source)?)
+    }
+
+    pub fn oracle_id(&self) -> DriftResult<OracleIdentifier> {
+        let oracle_source = self.get_oracle_source()?;
+        Ok((self.oracle, oracle_source))
+    }
+
+    pub fn get_last_available_amm_balance(&self) -> DriftResult<i128> {
+        let last_available_balance = self
+            .last_fee_pool_token_amount
+            .cast::<i128>()?
+            .safe_add(self.last_net_pnl_pool_token_amount)?;
+        Ok(last_available_balance)
+    }
+}
+
+#[zero_copy]
+#[derive(Default, Debug)]
+#[repr(C)]
+pub struct AmmCacheFixed {
+    pub bump: u8,
+    _pad: [u8; 3],
+    pub len: u32,
+}
+
+impl HasLen for AmmCacheFixed {
+    fn len(&self) -> u32 {
+        self.len
+    }
+}
+
+impl AmmCache {
+    pub fn space(num_markets: usize) -> usize {
+        8 + 8 + 4 + num_markets * CacheInfo::SIZE
+    }
+
+    pub fn validate(&self, state: &State) -> DriftResult<()> {
+        validate!(
+            self.cache.len() == state.number_of_markets as usize,
+            ErrorCode::DefaultError,
+            "Number of amm positions is different than number of markets"
+        )?;
+        Ok(())
+    }
+}
+
+impl_zero_copy_loader!(AmmCache, crate::id, AmmCacheFixed, CacheInfo);
+
+impl<'a> AccountZeroCopy<'a, CacheInfo, AmmCacheFixed> {
+    pub fn check_settle_staleness(&self, slot: u64, threshold_slot_diff: u64) -> DriftResult<()> {
```

Comment:
dont see this used any where but would if we plan on needed it, would be good to know the cu of looping through all the amms (~80?)